### ISSUE TITLE
PySM Design Notes

### ISSUE THREAD

**Issue by tskisner** [hpc4cmb/toast#148]
These are high-level notes about PySM integration.  Other, more specific issues should be created for the details.  At the highest level, we want a single toast.Operator class which loops over detectors and for each detector:

- Creates a distributed-ring-format map that has the same distribution as
  libsharp.
- Instantiate a pysm.Instrument object.
- Call pysm.observe for local pixels
- Call libsharp to do the harmonic transform, smoothing, and inverse transform
- Communicates (in some way) the distributed ring pixels to the process(es) 
  which are assigned to this detector
- Each process with data for this detector samples the pixels into a
  timestream.

There are several pieces to this.  Once we agree on these steps, I'll create
separate issues to track the pieces.

1.  Create a new class in tod/map/pixels.py that is similar to DistPixels.  Call it something like "DistRingPixels" or whatever you like.  The constructor would take:  communicator, nside, and nnz.  This class would use a toast.Cache() object for storing the data.  The size of the local data and the libsharp grid object would be computed with code like the "distribute_rings()" function in PR #145.

2.  Free function which takes a "DistRingPixels" map and smooths it (it is not a toast.Operator, since it just works with map domain data, not timestreams).  This function will already have access to the communicator and libsharp grid from the "DistRingPixels" object.  This function should go in a new file in toast/map/.  Maybe smoothing.py or harmonic.py or whatever you like.

3.  High-level OpSimPySM (like PR #137), but with these features/changes:
- constructor takes any properties that apply to all detectors and all observations.  Perhaps this is just the NSIDE value to use.
- exec() method loops over "observations".  Each observation could be from an entirely different telescope with different detectors and has a different group of processes working on it.  For each observation:
  - Get the group communicator from the tod class for this observation
  - Instantiate a DistRingPixels object on that communicator.
  - Loop over the global detector list (tod.detectors not tod.local_dets), calling pysm, smoothing, and redistribution / sampling.  So all processes will participate in this work, even for detectors that they do not have locally.

**Comment by jdborrill** [comment]
The general approach should also be extensible to other beam types -
elliptical, compact irregular (pixel domain), diffuse irregular
(libCONVIQT) - and their associated sky-sampling protocols.

Julian

**Comment by tskisner** [comment]
Yes.  I'll open another more detailed issue for just point 3 above, but briefly it would be like this in pseudocode:
```
pysm()
if symmetric:
    smoothing()
    redistribute()
    sample()
elif conviqt:
    # convert from distributed libsharp format to libconviqt format
    redistribute_conviqt()
    # libconviqt does the convolution and the sampling
    conviqt()
elif some_other_thing:
    blah()
```

**Comment by zonca** [comment]
I missed this note before implementing #152 , anyway I think I got it right based on your discussion.
@tskisner can you please tag me on future PySM related issues?

**Comment by zonca** [comment]
@tskisner I think it would help me understand the big picture if you can point me to one of the simplest pipelines where we will want to add PySM simulation capability.

**Comment by zonca** [comment]
`DistRings` handles all of the metadata for a ring-based distribution, this is independent of channel and observation. I would instead keep the actual distributed map outside of this object in a separate `toast.Cache`, so we can just re-use one single `DistRings` object.

**Comment by zonca** [comment]
See the ground pipeline because it reads from an input map: https://github.com/hpc4cmb/toast/blob/master/pipelines/toast_ground_sim.py#L583

**Comment by zonca** [comment]
For now we don't even need to shuffle the pixels to locally hit pixels and just do a AllReduce, then if we want to save memory we can do point-to-point.

**Comment by tskisner** [comment]
To summarize offline discussion, I think the current plan is:

1. Inside the exec() method, loop over observations and find the superset of all detectors that will ever be used.
2.  Instantiate a distributed ring map.

Loop over detectors:
1.  Call pysm to do the band integration.
2.  "Beam setup":
    - if symmetric, smooth map with beam and do an AllReduce so every process has a copy of the full map.  If this becomes too much for memory, Instantiate a DistPixels class (which requires a pass through the pointing) and then communicate pixel values.
    - else if using conviqt, read beam and construct all objects for this detector
3.  Loop over observations.  If the detector is not present, continue.
    - If symmetric, get pixel and I/Q/U weights from the tod cache and sample the map into a timestream
    - else if using conviqt, pass in pointing angles to the conviqt convolver, which returns timestreams.

**Comment by zonca** [comment]
implemented in #164, this issue can be closed
